### PR TITLE
state: links any missing binaries

### PR DESCRIPTION
**Pull request by quad**
Hermit ensures all binaries for a package have been linked.

This resolves crashes that occur as a result of adding [rust-analyzer][1].

[1]: https://github.com/cashapp/hermit-packages/pull/300

### CODE REVIEW

File: state/state_test.go [hunk 1]
```diff
@@ -77,6 +77,45 @@ func TestCacheAndUnpackHooksRunOnMutablePackage(t *testing.T) {
 	assert.Equal(t, os.FileMode(0500), info.Mode()&0777, info.Mode().String())
 }
 
+func TestUnpacksMissingBinaries(t *testing.T) {
```
Comment (alecthomas) [inline state/state_test.go:80]: Can you add an integration test to `integration/` too (or instead) please?

File: state/state.go [hunk 2]
```diff
@@ -407,8 +412,25 @@ func (s *State) isExtracted(p *manifest.Package) bool {
...
+	}
+
+	bins, err := p.ResolveBinaries()
+	if err != nil {
+		return false
+	}
+
+	binsDir := filepath.Join(s.binaryDir, p.Reference.String())
+
+	for _, bin := range bins {
```
Comment (alecthomas) [inline state/state.go:426]: This code is already in `linkBinaries()`, but also it shouldn't be in this function. I thought the plan was to just check if the symlinks existed then let the normal linking occur?

**Review by alecthomas — CHANGES_REQUESTED** [review]

File: state/state.go (hunk 2 above)
Comment (quad) [reply state/state.go:426]: oh shit thanks I misimplemented! 😅 I wrote this right after waking up.

File: state/state_test.go (hunk 1 above)
Comment (quad) [reply state/state_test.go:80]: Can do!

File: state/state.go (hunk 2 above)
Comment (quad) [reply state/state.go]: Fixed. Let me know if this implementation works for you. ❤️

File: state/state_test.go (hunk 1 above)
Comment (quad) [reply state/state_test.go]: How do I change a package definition or introspect state in the integration tests?

**Review by alecthomas — APPROVED** [review]

**Pull request merged**